<a href="https://colab.research.google.com/github/alexanderbandeiralira/CaseAilos-AlexanderLira-/blob/main/Case_Ailos_(AlexanderLira).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [31]:
from google.colab import drive
drive.mount('/content/drive', force_remount=True)

Mounted at /content/drive


# **Avaliação Ailos – Banco de Dados (Alexander Lira)**

### ***”In God we trust. All others must bring data.” - W. Edwards Deming***


# **Carregamento e pré processamento dos dados**

Nos dados, foram encontrados alguns problemas relacionados aos tipos das variáveis, como datas, CPFs e contas. Além disso, algumas variáveis monetárias estão em formatos diferentes do padrão que o Python consegue processar adequadamente para realizar funções e operações. Portanto, será necessário fazer algumas modificações nos tipos das variáveis para corrigir esses problemas.

Vamos agora proceder com as alterações necessárias para garantir que os dados estejam no formato adequado e assim possibilitar a realização das análises e operações desejadas.

In [32]:
import pandas as pd
import numpy as np
from scipy import stats
from IPython.display import display

In [33]:
base_carteira = pd.read_csv('/content/drive/MyDrive/CaseAilos/base_carteira.csv', encoding='latin-1', sep=';')
base_originacao = pd.read_csv('/content/drive/MyDrive/CaseAilos/base_originacao.csv', encoding='latin-1', sep=';')

In [34]:
base_carteira.head()

DTBASE COOPERATIVA  CPF   CONTA  CONTRATO   DSPRODUTO  DTCONTRATO  \
0  31/01/2021           A  154   49834    349384  Empréstimo  25/08/2020   
1  31/01/2021           A  186  398914    444074  Empréstimo  08/01/2021   
2  31/01/2021           A  430  549582    372569  Empréstimo  02/10/2020   
3  31/01/2021           A  512  217818    223696  Empréstimo  31/01/2020   
4  31/01/2021           A  754  546819    435007  Empréstimo  23/12/2020   

  VLCONTRATO SDDEVEDOR  QTDIASATRASO SDATRASO VLPROVISAO  
0   9.761,40  9.557,40           0.0     0,00      30,60  
1   4.314,60  4.365,60           0.0     0,00      81,60  
2   5.548,80  5.273,40           0.0     0,00      20,40  
3   3.111,00  2.539,80           0.0     0,00      10,20  
4     601,80    601,80          19.0   601,80      10,20

In [35]:
base_carteira.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44329 entries, 0 to 44328
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   DTBASE        44329 non-null  object 
 1   COOPERATIVA   44329 non-null  object 
 2   CPF           44329 non-null  int64  
 3   CONTA         44329 non-null  int64  
 4   CONTRATO      44329 non-null  int64  
 5   DSPRODUTO     44329 non-null  object 
 6   DTCONTRATO    44329 non-null  object 
 7   VLCONTRATO    44329 non-null  object 
 8   SDDEVEDOR     44329 non-null  object 
 9   QTDIASATRASO  44271 non-null  float64
 10  SDATRASO      44271 non-null  object 
 11  VLPROVISAO    44329 non-null  object 
dtypes: float64(1), int64(3), object(8)
memory usage: 4.1+ MB


In [36]:
base_originacao.head()

COOPERATIVA  CPF   CONTA  CONTRATO      DSPRODUTO  DTCONTRATO VLCONTRATO  \
0           A   32  496314    677366  Financiamento  13/10/2021   4.263,60   
1           A   38  624412    644724     Empréstimo  01/09/2021      81,60   
2           A  154   49834    349384     Empréstimo  25/08/2020   9.761,40   
3           A  186  398914    444074     Empréstimo  08/01/2021   4.314,60   
4           A  430  549582    372569     Empréstimo  02/10/2020   5.548,80   

  VLPARCELA  QTPARCELAS  
0    153,00        36.0  
1     10,20        10.0  
2    204,00        72.0  
3    163,20        36.0  
4    183,60        48.0

In [37]:
colunas_correcao = ['SDDEVEDOR', 'VLCONTRATO', 'SDATRASO', 'VLPROVISAO']

# Loop for para aplicar as correções em cada coluna
for coluna in colunas_correcao:
    base_carteira[coluna] = base_carteira[coluna].str.replace(r'\.', '').str.replace(r',', '.').astype(float)

colunas_correcao = ['VLCONTRATO', 'VLPARCELA']

# Loop for para aplicar as correções em cada coluna
for coluna in colunas_correcao:
    base_originacao[coluna] = base_originacao[coluna].str.replace(r'\.', '').str.replace(r',', '.').astype(float)


<ipython-input-37-9d0474c85b18>:5: FutureWarning: The default value of regex will change from True to False in a future version.
  base_carteira[coluna] = base_carteira[coluna].str.replace(r'\.', '').str.replace(r',', '.').astype(float)
<ipython-input-37-9d0474c85b18>:11: FutureWarning: The default value of regex will change from True to False in a future version.
  base_originacao[coluna] = base_originacao[coluna].str.replace(r'\.', '').str.replace(r',', '.').astype(float)


In [38]:
base_originacao.head()

COOPERATIVA  CPF   CONTA  CONTRATO      DSPRODUTO  DTCONTRATO  VLCONTRATO  \
0           A   32  496314    677366  Financiamento  13/10/2021      4263.6   
1           A   38  624412    644724     Empréstimo  01/09/2021        81.6   
2           A  154   49834    349384     Empréstimo  25/08/2020      9761.4   
3           A  186  398914    444074     Empréstimo  08/01/2021      4314.6   
4           A  430  549582    372569     Empréstimo  02/10/2020      5548.8   

   VLPARCELA  QTPARCELAS  
0      153.0        36.0  
1       10.2        10.0  
2      204.0        72.0  
3      163.2        36.0  
4      183.6        48.0

In [39]:
base_carteira.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44329 entries, 0 to 44328
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   DTBASE        44329 non-null  object 
 1   COOPERATIVA   44329 non-null  object 
 2   CPF           44329 non-null  int64  
 3   CONTA         44329 non-null  int64  
 4   CONTRATO      44329 non-null  int64  
 5   DSPRODUTO     44329 non-null  object 
 6   DTCONTRATO    44329 non-null  object 
 7   VLCONTRATO    44329 non-null  float64
 8   SDDEVEDOR     44329 non-null  float64
 9   QTDIASATRASO  44271 non-null  float64
 10  SDATRASO      44271 non-null  float64
 11  VLPROVISAO    44329 non-null  float64
dtypes: float64(5), int64(3), object(4)
memory usage: 4.1+ MB


In [40]:
base_carteira['DTBASE'] = pd.to_datetime(base_carteira['DTBASE'], format='%d/%m/%Y')
base_carteira['DTCONTRATO'] = pd.to_datetime(base_carteira['DTCONTRATO'], format='%d/%m/%Y')
base_carteira['CONTA'] = base_carteira['CONTA'].astype(str)
base_carteira['CONTRATO'] = base_carteira['CONTRATO'].astype(str)

In [41]:
base_originacao.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7217 entries, 0 to 7216
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   COOPERATIVA  7217 non-null   object 
 1   CPF          7217 non-null   int64  
 2   CONTA        7217 non-null   int64  
 3   CONTRATO     7217 non-null   int64  
 4   DSPRODUTO    7217 non-null   object 
 5   DTCONTRATO   7217 non-null   object 
 6   VLCONTRATO   7217 non-null   float64
 7   VLPARCELA    7212 non-null   float64
 8   QTPARCELAS   7212 non-null   float64
dtypes: float64(3), int64(3), object(3)
memory usage: 507.6+ KB


In [42]:
base_originacao['DTCONTRATO'] = pd.to_datetime(base_originacao['DTCONTRATO'], format='%d/%m/%Y')
base_originacao['CONTA'] = base_originacao['CONTA'].astype(str)
base_originacao['CONTRATO'] = base_originacao['CONTRATO'].astype(str)

# **Exercícios**

A partir daqui, serão abordadas as questões feitas na linguagem Python. A escolha desta linguagem se deve ao fato de que, ao final, também serão apresentados dois exemplos de como a questão 1-a pode ser resolvida usando consultas SQL e PySpark.

# **Questão 1 - a**

In [43]:
base_originacao.columns

Index(['COOPERATIVA', 'CPF', 'CONTA', 'CONTRATO', 'DSPRODUTO', 'DTCONTRATO',
       'VLCONTRATO', 'VLPARCELA', 'QTPARCELAS'],
      dtype='object')

In [44]:
# Calcular a quantidade de dias distintos em 'DTCONTRATO'
quantidade_dias_distintos = base_originacao['DTCONTRATO'].nunique()

# Calcular a soma dos 'VLCONTRATO' dos contratos originados
soma_vlcontrato = base_originacao['VLCONTRATO'].sum()

# Exibir os resultados
print(f"Quantidade de dias distintos em DTCONTRATO: {quantidade_dias_distintos}")
print(f"Soma dos VLCONTRATO dos contratos originados: {soma_vlcontrato:.2f}")

Quantidade de dias distintos em DTCONTRATO: 921
Soma dos VLCONTRATO dos contratos originados: 32816286.60


Na base "base_originacao.csv", foram encontrados 921 dias distintos em que foram realizados contratos (DTCONTRATO). Isso indica que ao longo do período registrado nessa base, houve 921 datas únicas em que novos contratos foram originados.

Além disso, a soma total dos valores dos contratos (VLCONTRATO) que foram originados na base é de 32.816.286,60. Esse valor representa o montante total de dinheiro envolvido em todos os contratos originados, ou seja, a soma de todos os valores de empréstimos ou financiamentos concedidos nesse período.

# **Questão 1 - b**

In [45]:
# Filtrar os dados para manter apenas as linhas com 'DTBASE' igual a '31/12/2021'
dados_31_12_2021 = base_carteira[base_carteira['DTBASE'] == '2021-12-31']

# Calcular a soma dos 'SDDEVEDOR'
soma_sddevedor = dados_31_12_2021['SDDEVEDOR'].sum()

# Calcular a média simples de 'QTDIASATRASO'
media_qtdiasatraso = dados_31_12_2021['QTDIASATRASO'].mean()

# Exibir os resultados
print(f"Soma dos SDDEVEDOR em 31/12/2021: {soma_sddevedor:.2f}")
print(f"Média simples de QTDIASATRASO em 31/12/2021: {media_qtdiasatraso:.2f}")

Soma dos SDDEVEDOR em 31/12/2021: 18183172.80
Média simples de QTDIASATRASO em 31/12/2021: 31.90


Em 31/12/2021, a soma total dos saldos devedores (SDDEVEDOR) de todos os contratos na base "base_carteira.csv" foi de 18.183.172,80. Isso representa o montante total de dinheiro em dívidas que os clientes têm com a instituição financeira nessa data específica.

Além disso, a média simples dos dias de atraso (QTDIASATRASO) em 31/12/2021 foi de 31,90 dias. Essa média indica que, em média, os contratos nessa data apresentavam um atraso de aproximadamente 31,90 dias em seus pagamentos.

# **Questão 2**

In [46]:
# Filtrar os dados para manter apenas a data de referência '30/09/2021'
dados_referencia = base_carteira[base_carteira['DTBASE'] == '2021-09-30']

# Filtrar os CPFs que possuem atraso maior que zero
dados_atraso_maior_que_zero = dados_referencia[dados_referencia['QTDIASATRASO'] > 0]

# Filtrar os dados apenas para a Cooperativa A
dados_cooperativa_A = dados_atraso_maior_que_zero[dados_atraso_maior_que_zero['COOPERATIVA'] == 'A']

# Ordenar os dados pelo saldo devedor em ordem decrescente
dados_cooperativa_A_ordenados = dados_cooperativa_A.sort_values(by='SDDEVEDOR', ascending=False)

# Obter os 10 CPFs com maiores Saldos Devedores
top_10_cpf_maior_saldo = dados_cooperativa_A_ordenados.head(10)

# Exibir a relação dos 10 CPFs com maiores Saldos Devedores
display(top_10_cpf_maior_saldo[['CPF', 'SDDEVEDOR']])

CPF  SDDEVEDOR
27033   80770    58425.6
29413  565660    40269.6
30185  714760    18421.2
27261  127329    17941.8
27443  164078    16697.4
28057  286984    16442.4
29727  629355    16411.8
27949  265230    15759.0
28685  409432    14137.2
28292  334457    14055.6

Essa lista pode ser útil para a cooperativa identificar os clientes com os maiores débitos em atraso, possibilitando ações específicas para recuperação desses créditos ou para ajustar as estratégias de cobrança e gestão de risco, a fim de reduzir a inadimplência e melhorar a saúde financeira da cooperativa.

# **Questão 3**

In [47]:
# Realizar um merge dos DataFrames base_carteira e base_originacao
base_carteira_merge = base_carteira.merge(base_originacao[['CPF', 'CONTRATO', 'QTPARCELAS']],
                                        on=['CPF', 'CONTRATO'], how='left')

# Criar uma lista vazia para armazenar os DataFrames de cada data base
data_frames_consolidados = []

# Agrupar o DataFrame 'base_carteira' por data base (DTBASE)
grouped_carteira = base_carteira_merge.groupby('DTBASE')

# Loop para processar cada data base e calcular as informações consolidadas
for data_base, grupo in grouped_carteira:
    # Quantidade de cooperativas distintas
    qtd_cooperativas = grupo['COOPERATIVA'].nunique()

    # Quantidade de CPFs distintos
    qtd_cpfs = grupo['CPF'].nunique()

    # Soma do saldo devedor
    saldo_devedor_total = round(grupo['SDDEVEDOR'].sum(), 1)

    # Média dos dias de atraso
    media_dias_atraso = round(grupo['QTDIASATRASO'].mean(), 1)

    # Menor data de originação dos contratos
    min_dtcontrato = pd.to_datetime(grupo['DTCONTRATO'], format='%d/%m/%Y').min()

    # Contrato com maior número de parcelas
    contrato_mais_parcelas = grupo.loc[grupo['QTPARCELAS'].idxmax(), 'CONTRATO']

    # Prazo médio ponderado dos contratos (valor da originação e quantidade de parcelas)
    prazo_medio_ponderado = round((grupo['VLCONTRATO'] * grupo['QTPARCELAS']).sum() / grupo['VLCONTRATO'].sum(), 2)

    # Taxa de contratos com atrasos > 90 dias
    taxa_atraso_90dias_contratos = round((grupo['QTDIASATRASO'] > 90).mean(),2)

    # Taxa de saldo devedor com atrasos > 90 dias
    taxa_atraso_90dias_saldo = round((grupo.loc[grupo['QTDIASATRASO'] > 90, 'SDDEVEDOR'].sum()) / saldo_devedor_total, 2)

    # Média da quantidade de parcelas
    media_qtd_parcelas = round(grupo['QTPARCELAS'].mean(), 2)

    # Criar um DataFrame com as informações da data base
    data_frame_consolidado = pd.DataFrame({'DTBASE': [data_base],
                                           'QTD_COOPERATIVAS': [qtd_cooperativas],
                                           'QTD_CPFS': [qtd_cpfs],
                                           'SALDO_DEVEDOR': [saldo_devedor_total],
                                           'MEDIA_DIAS_ATRASO': [media_dias_atraso],
                                           'MEDIA_QTD_PARCELAS': [media_qtd_parcelas],
                                           'MIN_DTCONTRATO': [min_dtcontrato],
                                           'CONTRATO_MAIS_PARCELAS': [contrato_mais_parcelas],
                                           'PRAZO_MEDIO_PONDERADO': [prazo_medio_ponderado],
                                           'TAXA_ATRASO_90DIAS_CONTRATOS': [taxa_atraso_90dias_contratos],
                                           'TAXA_ATRASO_90DIAS_SALDO': [taxa_atraso_90dias_saldo]})

    # Adicionar o DataFrame da data base à lista de DataFrames consolidados
    data_frames_consolidados.append(data_frame_consolidado)

# Concatenar todos os DataFrames consolidados em uma única tabela
tabela_consolidada = pd.concat(data_frames_consolidados, ignore_index=True)

# Ordenar a tabela consolidada pela DTBASE em ordem crescente
tabela_consolidada = tabela_consolidada.sort_values(by='DTBASE')

# Exibir a tabela consolidada
display(tabela_consolidada)

DTBASE  QTD_COOPERATIVAS  QTD_CPFS  SALDO_DEVEDOR  MEDIA_DIAS_ATRASO  \
0  2021-01-31                 5      3333     13215691.2               22.9   
1  2021-02-28                 4      3334     13058845.8               21.8   
2  2021-03-31                 5      3599     14389670.4               22.0   
3  2021-04-30                 5      3704     15181006.8               22.0   
4  2021-05-31                 4       571      2399193.0               28.4   
5  2021-06-30                 5      3918     15556938.0               24.6   
6  2021-07-31                 5      4011     16769962.8               25.0   
7  2021-08-31                 5      4119     17236592.4               26.3   
8  2021-09-30                 5      4313     17437828.2               27.2   
9  2021-10-31                 5      4395     17903427.6               30.4   
10 2021-11-30                 3      4340     17600028.6               32.4   
11 2021-12-31                 5      4692     18183172.8               31.9   

    MEDIA_QTD_PARCELAS MIN_DTCONTRATO CONTRATO_MAIS_PARCELAS  \
0                37.00     2013-01-24                 301340   
1                37.16     2013-01-24                 474081   
2                37.10     2013-01-24                 474081   
3                37.23     2013-01-24                 498576   
4                37.71     2014-07-15                  30084   
5                37.00     2013-01-24                 474081   
6                37.41     2013-01-24                 498576   
7                38.34     2013-01-24                 498576   
8                37.88     2013-01-24                 498576   
9                38.04     2013-01-24                 498576   
10               37.70     2013-01-24                 498576   
11               37.50     2013-01-24                 549274   

    PRAZO_MEDIO_PONDERADO  TAXA_ATRASO_90DIAS_CONTRATOS  \
0                   61.90                          0.10   
1                   60.95                          0.09   
2                   63.01                          0.09   
3                   64.66                          0.09   
4                   61.57                          0.13   
5                   62.01                          0.10   
6                   69.34                          0.11   
7                   69.31                          0.11   
8                   69.05                          0.12   
9                   69.32                          0.13   
10                  70.34                          0.13   
11                  67.01                          0.13   

    TAXA_ATRASO_90DIAS_SALDO  
0                       0.04  
1                       0.04  
2                       0.04  
3                       0.04  
4                       0.06  
5                       0.04  
6                       0.04  
7                       0.05  
8                       0.05  
9                       0.05  
10                      0.06  
11                      0.06

Na maior parte dos meses o contrato com maior numero de parcelas é o 498576.

In [48]:
tabela_consolidada.describe()

QTD_COOPERATIVAS     QTD_CPFS  SALDO_DEVEDOR  MEDIA_DIAS_ATRASO  \
count         12.000000    12.000000   1.200000e+01          12.000000   
mean           4.666667  3694.083333   1.491103e+07          26.241667   
std            0.651339  1072.013692   4.324911e+06           3.860395   
min            3.000000   571.000000   2.399193e+06          21.800000   
25%            4.750000  3532.750000   1.409618e+07          22.675000   
50%            5.000000  3964.500000   1.616345e+07          25.650000   
75%            5.000000  4319.750000   1.747838e+07          28.900000   
max            5.000000  4692.000000   1.818317e+07          32.400000   

       MEDIA_QTD_PARCELAS  PRAZO_MEDIO_PONDERADO  \
count           12.000000              12.000000   
mean            37.505833              65.705833   
std              0.435044               3.690862   
min             37.000000              60.950000   
25%             37.145000              61.982500   
50%             37.455000              65.835000   
75%             37.752500              69.312500   
max             38.340000              70.340000   

       TAXA_ATRASO_90DIAS_CONTRATOS  TAXA_ATRASO_90DIAS_SALDO  
count                     12.000000                  12.00000  
mean                       0.110833                   0.04750  
std                        0.016765                   0.00866  
min                        0.090000                   0.04000  
25%                        0.097500                   0.04000  
50%                        0.110000                   0.04500  
75%                        0.130000                   0.05250  
max                        0.130000                   0.06000

Nessa tabela, podemos observar algumas informações relevantes sobre as métricas consolidadas para a carteira de contratos:

- **QTD_COOPERATIVAS:** A média indica que foram analisadas cerca de 4.67 cooperativas durante o período.

- **QTD_CPFS:** A média de cerca de 3694 CPFs mostra quantos clientes individuais estão representados na carteira de contratos.

- **SALDO_DEVEDOR:** A média do saldo devedor é de aproximadamente 1.491103e+07	 representando o valor médio total dos débitos em reais.

- **MEDIA_DIAS_ATRASO:** A média de cerca de 26.24 dias de atraso representa o tempo médio em dias que os pagamentos estão atrasados.

- **MEDIA_QTD_PARCELAS:** A média de aproximadamente 37.51 parcelas representa o número médio de parcelas em cada contrato.

- **PRAZO_MEDIO_PONDERADO:** A média de cerca de 65.71 indica o prazo médio ponderado, levando em consideração tanto o valor do contrato quanto o número de parcelas.

- **TAXA_ATRASO_90DIAS_CONTRATOS:** A taxa média de atraso de aproximadamente 11.08% indica a porcentagem de contratos que estão atrasados por mais de 90 dias.

- **TAXA_ATRASO_90DIAS_SALDO:** A taxa média de atraso de aproximadamente 4.75% indica a porcentagem média do saldo devedor que está atrasado por mais de 90 dias.

Essas métricas fornecem insights que podem ser relavante a depender do problema de negócio que se deseja solucionar.

Obs: Aqui utilizamos a média porque aparentemente, pela distribuição dos quantis, não temos outliers que tragam um viés para nossa média e nos faça utilizar a mediana como medida resumo dos nossos dados.

# **Questão 4**

Nessa questão vamos ter uma visão mensal (DTBASE), para o atributo QTPARCELAS, da Moda, e para o atributo SDDEVEDOR a Média e o Intervalo de Confiança para a média com nível de 99%.

O intervalo de confiança é uma medida estatística importante que nos ajuda a quantificar a incerteza em nossas estimativas de parâmetros populacionais com base em amostras. É um intervalo no qual temos um certo nível de confiança de que o verdadeiro valor do parâmetro está contido. Um nível de confiança de 99% significa que, em média, se repetirmos a amostragem e a análise muitas vezes, em cerca de 99% dos casos, o intervalo conterá o verdadeiro valor do parâmetro populacional. Ele vai nos fornecer uma medida de quão precisas são essas inferências, considerando a variabilidade dos dados amostrais.

In [49]:
# Realizar o merge dos DataFrames base_carteira e base_originacao
base_carteira_merge = base_carteira.merge(base_originacao[['CPF', 'CONTRATO', 'QTPARCELAS']],
                                    on=['CPF', 'CONTRATO'], how='left')


# Criar uma lista vazia para armazenar os DataFrames de cada mês
data_frames_consolidados = []

# Agrupar o DataFrame 'base_carteira' por data base (DTBASE)
grouped_carteira = base_carteira_merge.groupby('DTBASE')

# Loop para processar cada mês e calcular as informações consolidadas
for data_base, grupo in grouped_carteira:
    # Calcular a Moda para a coluna QTPARCELAS
    moda_qtd_parcelas = grupo['QTPARCELAS'].mode().values[0]

    # Calcular a Média e o Intervalo de Confiança a 99% para a coluna SDDEVEDOR
    media_sddevedor = round(grupo['SDDEVEDOR'].astype(float).mean(),2)
    intervalo_confianca_sddevedor = stats.t.interval(0.99, len(grupo['SDDEVEDOR']) - 1, loc=media_sddevedor, scale=stats.sem(grupo['SDDEVEDOR'].astype(float)))

    # Criar um DataFrame com as informações do mês
    data_frame_consolidado = pd.DataFrame({'DTBASE': [data_base],
                                           'MODA_QTD_PARCELAS': [moda_qtd_parcelas],
                                           'MEDIA_SDDEVEDOR': [media_sddevedor],
                                           'IC_99_INFERIOR_SDDEVEDOR': [round(intervalo_confianca_sddevedor[0], 2)],
                                           'IC_99_SUPERIOR_SDDEVEDOR': [round(intervalo_confianca_sddevedor[1], 2)]})

    # Adicionar o DataFrame do mês à lista de DataFrames consolidados
    data_frames_consolidados.append(data_frame_consolidado)

# Concatenar todos os DataFrames consolidados em uma única tabela
tabela_consolidada = pd.concat(data_frames_consolidados, ignore_index=True)

# Exibir a tabela consolidada
display(tabela_consolidada)


DTBASE  MODA_QTD_PARCELAS  MEDIA_SDDEVEDOR  IC_99_INFERIOR_SDDEVEDOR  \
0  2021-01-31               48.0          3965.10                   3628.35   
1  2021-02-28               48.0          3916.87                   3614.49   
2  2021-03-31               48.0          3998.24                   3678.98   
3  2021-04-30               48.0          4098.54                   3761.39   
4  2021-05-31               48.0          4201.74                   3299.50   
5  2021-06-30               48.0          3970.63                   3673.42   
6  2021-07-31               48.0          4180.99                   3836.30   
7  2021-08-31               48.0          4184.65                   3846.97   
8  2021-09-30               48.0          4043.09                   3719.63   
9  2021-10-31               48.0          4073.59                   3725.67   
10 2021-11-30               48.0          4055.31                   3698.84   
11 2021-12-31               48.0          3875.36                   3538.27   

    IC_99_SUPERIOR_SDDEVEDOR  
0                    4301.85  
1                    4219.25  
2                    4317.50  
3                    4435.69  
4                    5103.98  
5                    4267.84  
6                    4525.68  
7                    4522.33  
8                    4366.55  
9                    4421.51  
10                   4411.78  
11                   4212.45

Ao analisar os resultados apresentados na tabela, podemos identificar algumas tendências e informações importantes sobre os dados consolidados ao longo do ano de 2021:

**Tendências no Saldo Devedor Médio:** Podemos observar que o saldo devedor médio apresentou variações ao longo do ano. Iniciando em janeiro com um valor de aproximadamente 3965.10 e alcançando um pico em julho, com um valor próximo a 4180.99. A partir de então, houve uma leve queda em agosto e setembro, seguida de um aumento em outubro e novembro, e uma leve queda em dezembro, chegando a um valor próximo a 3875.36.

**Estabilidade na Moda da Quantidade de Parcelas:** A coluna "MODA_QTD_PARCELAS" mostra que a moda da quantidade de parcelas dos contratos é constante, permanecendo em 48 parcelas em todos os meses.

**Importância do Intervalo de Confiança:** O intervalo de confiança é essencial para avaliar a precisão das estimativas. Quanto mais estreito o intervalo de confiança, mais precisa é a estimativa da média. Nesse contexto, um intervalo de confiança mais amplo indica uma maior incerteza e variabilidade nos dados observados.

# **Questão 5**

In [50]:
# Realizar o merge dos dataframes base_carteira e base_originacao
merged_df = base_carteira.merge(base_originacao[['COOPERATIVA', 'CPF', 'CONTA', 'CONTRATO', 'DSPRODUTO', 'DTCONTRATO', 'VLCONTRATO']],
                                on=['COOPERATIVA', 'CPF', 'CONTA', 'CONTRATO', 'DSPRODUTO', 'DTCONTRATO', 'VLCONTRATO'],
                                how='inner')

# Filtrar os contratos em 31/jul/2021 e 31/ago/2021
contratos_31_jul = merged_df[merged_df['DTBASE'] == '2021-07-31']
contratos_31_ago = merged_df[merged_df['DTBASE'] == '2021-08-31']

# Identificar os contratos em 31/jul/2021 que não estão em 31/ago/2021
contratos_nao_em_31_ago = contratos_31_jul[~contratos_31_jul['CONTRATO'].isin(contratos_31_ago['CONTRATO'])]

# Agrupar por cooperativa e contar a quantidade de contratos para cada uma
volume_contratos_por_cooperativa = contratos_nao_em_31_ago.groupby('COOPERATIVA').size().reset_index(name='VOLUME_CONTRATOS')

# Exibir o resultado
display(volume_contratos_por_cooperativa)


COOPERATIVA  VOLUME_CONTRATOS
0           A               299
1           B                 5
2           C                10
3           D                14
4           E                25

A cooperativa A lidera com um volume significativamente maior de 299 contratos, enquanto as cooperativas B, C, D e E possuem volumes menores, com 5, 10, 14 e 25 contratos, respectivamente.

# **Questão 6**

In [51]:
# Realizar o merge dos dataframes para obter as informações consolidadas
merged_df = base_carteira.merge(base_originacao[['COOPERATIVA', 'CPF', 'CONTA', 'CONTRATO', 'DSPRODUTO', 'DTCONTRATO', 'VLCONTRATO', 'QTPARCELAS']],
                                on=['COOPERATIVA', 'CPF', 'CONTA', 'CONTRATO', 'DSPRODUTO', 'DTCONTRATO', 'VLCONTRATO'],
                                how='inner')

# Filtrar os contratos pertencentes à cooperativa B
cooperativa_b = merged_df[merged_df['COOPERATIVA'] == 'B']

# Calcular as principais métricas para a cooperativa B
qtd_contratos = cooperativa_b['CONTRATO'].nunique()
saldo_devedor_medio = cooperativa_b['SDDEVEDOR'].mean()
media_dias_atraso = cooperativa_b['QTDIASATRASO'].mean()
qtd_contratos_atrasados = (cooperativa_b['QTDIASATRASO'] > 0).sum()
taxa_atraso_90dias = (cooperativa_b['QTDIASATRASO'] > 90).mean()
prazo_medio_ponderado = (cooperativa_b['VLCONTRATO'] * cooperativa_b['QTPARCELAS']).sum() / cooperativa_b['VLCONTRATO'].sum()

# Apresentar os resultados em um sumário executivo
sumario_executivo = pd.DataFrame({
    'Cooperativa': ['B'],
    'Quantidade de Contratos': [qtd_contratos],
    'Saldo Devedor Médio': [saldo_devedor_medio],
    'Média de Dias de Atraso': [media_dias_atraso],
    'Quantidade de Contratos Atrasados': [qtd_contratos_atrasados],
    'Taxa de Atraso > 90 dias': [taxa_atraso_90dias],
    'Prazo Médio Ponderado': [prazo_medio_ponderado]
})


display(sumario_executivo)

Cooperativa  Quantidade de Contratos  Saldo Devedor Médio  \
0           B                      164          5032.695167   

   Média de Dias de Atraso  Quantidade de Contratos Atrasados  \
0                41.697674                                371   

   Taxa de Atraso > 90 dias  Prazo Médio Ponderado  
0                  0.178439              51.429051

Os dados apresentam informações consolidadas para a cooperativa "B" em relação aos seus contratos.

- Quantidade de Contratos: A cooperativa "B" possui 164 contratos em sua carteira.

- Saldo Devedor Médio: O saldo devedor médio dos contratos é de aproximadamente 5032.70 unidades monetárias.

- Média de Dias de Atraso: Os contratos apresentam um atraso médio de aproximadamente 41.70 dias no pagamento de parcelas.

- Quantidade de Contratos Atrasados: Há um total de 371 contratos com atraso no pagamento das parcelas.

- Taxa de Atraso > 90 dias: Cerca de 17.84% dos contratos estão com atraso superior a 90 dias.

- Prazo Médio Ponderado: O prazo médio ponderado dos contratos é de aproximadamente 51.43 unidades monetárias.

# **Extra**

## **Possibilidade de análise via SQL ou Pysparks do problema 1-a(Extra)**

### **SQL (Problema 1-a)**

In [52]:
import sqlite3

In [53]:
# Criar uma conexão com o banco de dados
conn = sqlite3.connect('bancodedados.db')

# Inserir os dados do DataFrame na tabela 'base_originacao' no banco de dados
base_originacao.to_sql('base_originacao', conn, if_exists='replace', index=False)

7217

In [54]:
# Consulta SQL para calcular a quantidade de dias distintos em 'DTCONTRATO'
query_quantidade_dias_distintos = """
SELECT COUNT(DISTINCT DTCONTRATO) AS quantidade_dias_distintos
FROM base_originacao;
"""

# Consulta SQL para calcular a soma dos 'VLCONTRATO' dos contratos originados
query_soma_vlcontrato = """
SELECT SUM(VLCONTRATO) AS soma_vlcontrato
FROM base_originacao;
"""

# Executar as consultas SQL e armazenar os resultados em DataFrames
df_quantidade_dias = pd.read_sql_query(query_quantidade_dias_distintos, conn)
df_soma_vlcontrato = pd.read_sql_query(query_soma_vlcontrato, conn)

# Fechar a conexão com o banco de dados para otimizar a velocidade
conn.close()

# Exibir os resultados
quantidade_dias_distintos = df_quantidade_dias['quantidade_dias_distintos'][0]
soma_vlcontrato = df_soma_vlcontrato['soma_vlcontrato'][0]

print(f"Quantidade de dias distintos em DTCONTRATO: {quantidade_dias_distintos}")
print(f"Soma dos VLCONTRATO dos contratos originados: {soma_vlcontrato:.2f}")


Quantidade de dias distintos em DTCONTRATO: 921
Soma dos VLCONTRATO dos contratos originados: 32816286.60


### **PySparks (Problema 1-a)**

In [55]:
#!pip install pyspark

In [56]:
from pyspark.sql import SparkSession
from pyspark.sql.functions import col, translate, countDistinct
from pyspark.sql import functions as F

In [57]:
# Inicializar a sessão do Spark
spark = SparkSession.builder.appName("Exemplo Spark").getOrCreate()

# Carregar o arquivo CSV em um DataFrame, especificando o separador como ';'
base_originacao_df = spark.createDataFrame(base_originacao)


# Calcular a quantidade de dias distintos em 'DTCONTRATO'
quantidade_dias_distintos = base_originacao_df.select(countDistinct("DTCONTRATO")).collect()[0][0]

# Calcular a soma dos 'VLCONTRATO' dos contratos originados
soma_vlcontrato = base_originacao_df.select(F.sum("VLCONTRATO")).collect()[0][0]

# Exibir os resultados
print(f"Quantidade de dias distintos em DTCONTRATO: {quantidade_dias_distintos}")
print(f"Soma dos VLCONTRATO dos contratos originados: {soma_vlcontrato:.2f}")

Quantidade de dias distintos em DTCONTRATO: 921
Soma dos VLCONTRATO dos contratos originados: 32816286.60
